haskell/test/SMT/Model/BaseDAO/Types.hs

```diff
@@ -0,0 +1,269 @@
+  ) where
+
+import Universum
+
+import Control.Monad.Except (MonadError, throwError)
+import Data.Typeable (cast)
+import qualified Data.Map as M
+import Fmt
+
+import Michelson.Interpret
+import Michelson.Text
+import qualified Michelson.Typed as T
+import Michelson.Typed.Value
+import Lorentz hiding (not, cast, get)
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+import Tezos.Address (ContractHash(..))
+import Tezos.Core
+
+import Ligo.BaseDAO.Types
+
+-- | Transformer used in haskell implementation of BaseDAO
+newtype ModelT a = ModelT
+  { unModelT :: (ExceptT ModelError $ State ModelState) a
+  } deriving newtype (Functor, Applicative, Monad, MonadState ModelState, MonadError ModelError)
+
+-- | A type that keep track of blockchain state
+data ModelState = ModelState
+  { msFullStorage :: FullStorage
+
+  , msLevel :: Natural
+  , msChainId :: ChainId
+  , msSelfAddress :: Address
+  , msGovernanceTokenAddress :: Address
+  , msContracts :: Map Address SimpleContractType
+  -- ^ Keep track of originated contract in the blockchain.
+  -- This only
+  } deriving stock (Eq, Generic, Show)
+
+runModelT :: ModelT a -> ModelState -> (Either ModelError ModelState)
+runModelT (ModelT model) initState =
+  runExceptT model
+    & (`runState` initState)
+    & (\(result, updatedMs) -> case result of
+          Left err -> Left err
+          Right _ -> Right updatedMs
+      )
+
+execOperations :: [SimpleOperation] -> ModelT ()
+execOperations ops = do
+  contracts <- get <&> msContracts
+  let updatedContracts =
+        foldl'
+          (\contractMap op ->
+              case op of
+                FA2TransferOperation addr param tez ->
+                  case M.lookup (addr) contractMap of
+                    Just (SimpleFA2ContractType sfc) ->
+                      let updatedSc = sfc
+                            { sfcStorage = param : (sfc & sfcStorage)
+                            , sfcMutez = unsafeAddMutez tez (sfc & sfcMutez)
+                            }
+                      in M.insert addr (SimpleFA2ContractType updatedSc) contractMap
+                    _ ->
+                      contractMap
+                OtherOperation addr param tez ->
+                  case M.lookup addr contractMap of
+                    Just (OtherContractType oc) ->
+                      let updatedSc = oc
+                            { ocStorage = param : (oc & ocStorage)
+                            , ocMutez = unsafeAddMutez tez (oc & ocMutez)
+                            }
+                      in M.insert addr (OtherContractType updatedSc) contractMap
+                    _ ->
+                      contractMap
+          )
+          (contracts)
+          (ops)
+  modify $ \ms -> ms { msContracts = updatedContracts }
+
+getStore :: ModelT Storage
+getStore = get <&> msFullStorage <&> fsStorage
+
+getConfig :: ModelT Config
+getConfig = get <&> msFullStorage <&> fsConfig
+
+modifyStore :: (Storage -> Storage) -> ModelT ()
```

Comment:
I suggest you to change the type of this to
```suggestion
modifyStore :: (Storage -> ModelT Storage) -> ModelT ()
```
many functions/"entrypoints" could more easily be rewritten inside a single `modifyStore` call.

Reply:
Updated.